### ISSUE TITLE
Imprecise type inference with complex generic + type[]

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#3576]
### Summary

On this sample ty fails to infer the expected precise type (`Sequence[str]`):
```python
from collections.abc import Sequence
from typing import Generic, TypeVar, reveal_type


T = TypeVar("T")


class Option(Generic[T]):
    pass


class StringOptions(Option[Sequence[str]]):
    pass


class Options:
    def get_value_for(self, option: type[Option[T]]) -> T:
        raise NotImplementedError


def check(options: Options) -> None:
    values = reveal_type(options.get_value_for(StringOptions)) # Unknown
    values.nonexistent() # not caught
```

Possibly similar to #2521 or #3469 but seems distinct.

**Comment by carljm** [comment]
This is a missing case in `infer_map_impl` in the old constraint solver; we only infer through `type[...]` if it's directly a `type[T]`, not if it's a `type[Something[T]]`. We should see if we can fix this by having this case (and ideally, as general a class of cases as possible) defer to constraint-set assignability instead of being special-cased in `infer_map_impl`.
